rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -288,6 +288,11 @@ class Publisher : public PublisherBase
     return message_seq;
   }
 
+  /// Copy of original options passed during construction.
+  /**
+   * It is important to save a copy of this so that the rmw payload which it
+   * may contain is kept alive for the duration of the publisher.
+   */
```

Comment:
I tried what you suggested first.

Reply:
You're currently copying the options when the `rmw_publisher` and `rmw_subscription` are created, so I don't understand why you also need to keep a copy in `rclcpp` (in the case of the `publisher`, a copy is needed because it's later used in [rclcpp code](https://github.com/ros2/rclcpp/blob/9b4f04927759b1e33f6b2d365fbed5cedb4e883b/rclcpp/include/rclcpp/publisher.hpp#L119)):
https://github.com/ros2/rmw_fastrtps/blob/b4200db001a642b86856705ada799fc00ee5395c/rmw_fastrtps_cpp/src/rmw_subscription.cpp#L191
https://github.com/ros2/rmw_fastrtps/blob/b4200db001a642b86856705ada799fc00ee5395c/rmw_fastrtps_cpp/src/rmw_publisher.cpp#L216

Reply:
> You're currently copying the options when the `rmw_publisher` and `rmw_subscription` are created, so I don't understand why you also need to keep a copy in `rclcpp`

That copy is just in case it is ever needed for reference from above the rmw layer in the future, e.g. to answer the question (from `rcl` or `rclcpp`) "was the subscription created with ignore local publications set to true?". Copying it there has nothing to do with managing the lifetime of the payload or destroying the payload when possible.

The rmw implementation cannot destruct this class safely, in my opinion. Therefore it needs to be kept around in `rclcpp`, so that it can be destroyed in C++ (above the rmw layer). I don't think it's a good idea for the rmw implementation to destroy the object in case the user has further subclassed it or just because you'd like to avoid `new` in one executable/library and then `delete` in a different library.

Reply:
> I don't think it's a good idea for the rmw implementation to destroy the object in case the user has further subclassed it or just because you'd like to avoid new in one executable/library and then delete in a different library.

Yes, that's also not nice.
I now understand why you are keeping the copy, and I don't have a better way of doing it, so this is fine.